README.md
[![Pub.dev package](https://img.shields.io/badge/pub.dev-code__scan-blue)](https://pub.dev/packages/code_scan)
[![GitHub repository](https://img.shields.io/badge/GitHub-CodeScanner--flutter-blue?logo=github)](https://github.com/DrafaKiller/CodeScanner-flutter)

# Code Scanner

A flexible code scanner for QR codes, barcodes and many others. Using [Google's ML Kit](https://developers.google.com/ml-kit/vision/barcode-scanning). Use it as a Widget with a camera or use the methods provided, with a camera controller.

## Features

* Scan Linear and 2D formats: QR Code, Barcode, ...
* Widget with integrated camera
* Listen for callbacks with every code scanned
* Choose which formats to scan
* Overlay the camera preview with a custom view
* Camera lifecycle

## Scannable Formats

* Aztec
* Codabar
* Code 39
* Code 93
* Code 128
* Data Matrix
* EAN-8
* EAN-13
* ITF
* PDF417
* QR Code
* UPC-A
* UPC-E

## Getting started

Install it using pub:
```
flutter pub add code_scan
```

And import the package:
```dart
import 'package:code_scan/code_scan.dart';
```

## Usage

```dart
CodeScanner(
    onScan: (code, details, controller) => setState(() => this.code = code),
    onScanAll: (codes, controller) => print('Codes: ' + codes.map((code) => code.rawValue).toString()),
    formats: [ BarcodeFormat.qrCode ],
    once: true,
)
``` 

Add callbacks for events:
```dart
CodeScanner(
    onScan: (code, details, controller) => ...,
    onScanAll: (codes, controller) => ...,
    onCreated: (controller) => ...,
)
```

Set `loading` and `overlay` widgets, although you can use the default overlay:
```dart
CodeScanner(
    loading: Center(child: CircularProgressIndicator()),
    overlay: Center(child: Text('Scanning...')),
)
```

Choose how the widget should react:
```dart
CodeScanner(
    direction: CameraLensDirection.back,
    resolution: ResolutionPreset.medium,
    formats: const [ BarcodeFormat.all ],
    scanInterval: const Duration(seconds: 1),
    aspectRatio: 480 / 720, // height / width
    once: false,
)
```

The default behavior of the widget is to fill the screen or parent widget, but you can choose the aspect ratio.

You can use the camera view with your camera controller, without having the scanning widget:
```dart
final cameraController = CameraController();

CodeScannerCameraView(
    controller: cameraController,
)
```

### Caution

If you use your camera controller, it's recommended that you be responsible to control camera resources.

Manage it yourself using [the lifecycle](https://pub.dev/packages/camera#handling-lifecycle-states) or make sure it's already being managed by some other widget.

## Methods provided

Create a camera listener and plug it with any camera controller, to scan for codes:
```dart
final cameraController = CameraController();

final listener = CodeScannerCameraListener(
    cameraController,
    
    formats: const [ BarcodeFormat.all ],
    interval: const Duration(milliseconds: 500),
    once: false,
    
    onScan: (code, details, controller) => ...,
    onScanAll: (codes, controller) => ...,
);
```

Stop the listener whenever you want:
```dart
listener.stop();
```

## GitHub

The package code is available on Github: [Flutter - CodeScanner](https://github.com/DrafaKiller/CodeScanner-flutter)


CHANGELOG.md
## 0.4.10

Fixed:
- Pub.dev format problem
- Updating CHANGELOG.md correctly 

## 0.4.9

Changed:
- Updated dependencies, google_mlkit_barcode_scanning to `0.6.0`

## 0.4.8

Changed:
- Updated dependencies, google_mlkit_barcode_scanning to `0.5.0`

## 0.4.7

Removed:
- `README.md` warning about the pub.dev package platform compatibility mislabeling

## 0.4.6

Fixed:
- Issue tracker in `pubspec.yaml`

## 0.4.5

Added:
- `README.md` warning about the pub.dev package platform compatibility mislabeling

Fixed:
- Fake lint error, again

## 0.4.4 

Fixed:
- Missing `flutter` to `pubspec.yaml`, to fix pub.dev's platform compatibilities. *(Did not work)*

## 0.4.3

Fixed:
- Flutter version to `>=2.10.0` in `pubspec.yaml`, hopefully fixing pub.dev's platform compatibilities. *(Did not work)*
- **CodeScannerCameraListener** `.dispose()` method now does the same as `.stop()` to make sure no more data passes through the listener, into the callbacks.

## 0.4.2

Added:
- `lints` developer dependency

Fixed:
- Lints error: Unnecessary `dart:typed_data` import

## 0.4.1

Fixed:
- Upgraded `google_mlkit_barcode_scanning` dependency version from `0.3.0` to `0.4.0`.

## 0.4.0

Added:
- Exported used interfaces `ResolutionPreset` and `BarcodeFormat`
- `onError` callback

Changed:
- Default `ResolutionPreset` from `medium` to `high`, because `medium` may not be supported on all devices ([CodeScanner-flutter/issues/1](https://github.com/DrafaKiller/CodeScanner-flutter/issues/1))

## 0.3.3

Fixed:
- "Memory leak" right before the app closes

## 0.3.2

Changed:
- `onAccessDenied` return value is optional, default is `false`

## 0.3.1

Added:
- `shield.io` badges

Fixed:
- ChangeLog typos

## 0.3.0

Added:
- `onAccessDenied` callback

Fixed:
- `CameraController` created in the Lifecycle would not have a `CodeScannerCameraListener`
- Scanner listener wouldn't be disposed of in the lifecycle


## 0.2.0

Added:
- Lifecycle states handling, on `CameraController` created internally
- Documentation for `CodeScanner` constructor
- Readme warning about the `CameraController` lifecycle

Changed:
- ChangeLog's format
- Package description

## 0.1.0

Changed:
- Callback hands `listener` instead of `controller`

## 0.0.2

Changed:
- Dart SDK version

## 0.0.1

Initial release: CodeScanner


lib/camera_view.dart
import 'package:camera/camera.dart';
import 'package:code_scan/camera_overlay.dart';
import 'package:flutter/widgets.dart';

/// Widget to show a camera with an overlay, this widget tries to expand.
class CodeScannerCameraView extends StatelessWidget {
  final CameraController controller;
  final Widget? overlay;
  final double? aspectRatio;

  const CodeScannerCameraView({
    super.key,
    required this.controller,
    this.overlay,
    this.aspectRatio,
  });
  
  @override
  Widget build(BuildContext context) {
    final cameraAspectRatio = controller.value.aspectRatio;
    return LayoutBuilder(
      builder: (context, constraints) => Stack(
        children: [
          SizedBox(
            width: constraints.maxWidth,
            height: constraints.maxHeight,
            child: FittedBox(
              fit: BoxFit.cover,
              clipBehavior: Clip.hardEdge,
              child: SizedBox(
                width: constraints.maxWidth,
                height: constraints.maxWidth * cameraAspectRatio,
                child: CameraPreview(controller),
              ),
            ),
          ),
          overlay ?? CodeScannerOverlay(constraints.maxWidth, constraints.minHeight),
        ],
      ),
    );
  }
}

lib/utils/image.dart
import 'dart:typed_data';

import 'package:camera/camera.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:google_mlkit_barcode_scanning/google_mlkit_barcode_scanning.dart';

extension CameraImageToInputImage on CameraImage {
  Uint8List getBytes() {
    final allBytes = WriteBuffer();
    for (final Plane plane in planes) {
      allBytes.putUint8List(plane.bytes);
    }
    return allBytes.done().buffer.asUint8List();
  }  

  InputImage toInputImage(CameraDescription camera) {
    final inputImageData = InputImageData(
      size: Size(width.toDouble(), height.toDouble()),
      imageRotation: InputImageRotationValue.fromRawValue(camera.sensorOrientation) ?? InputImageRotation.rotation0deg,
      inputImageFormat: InputImageFormatValue.fromRawValue(format.raw) ?? InputImageFormat.nv21,
      planeData: planes.map((Plane plane) => InputImagePlaneMetadata(
        bytesPerRow: plane.bytesPerRow,
        height: plane.height,
        width: plane.width,
      )).toList(),
    );
    return InputImage.fromBytes(bytes: getBytes(), inputImageData: inputImageData);
  }
}

class CameraImageData {
  final CameraImage image;
  final CameraDescription description;
  CameraImageData(this.image, this.description);
}

/*
TODO: -= Image Crop =- (To prevent the scanner to scan too many pixels)

extension InputImageCrop on InputImage {
  InputImage crop(int x, int y, int width, int height) {
    final croppedData = InputImageData(
      size: inputImageData!.size,
      imageRotation: inputImageData!.imageRotation,
      inputImageFormat: inputImageData!.inputImageFormat,
      planeData: inputImageData!.planeData!.map((InputImagePlaneMetadata planeData) => InputImagePlaneMetadata(
        bytesPerRow: width * planeData.bytesPerRow ~/ planeData.width!,
        height: height,
        width: width,
      )).toList(),
    );
    
    final croppedBytes = Uint8List.fromList(bytes!.map((byte) => byte).toList());
    
    return InputImage.fromBytes(
      bytes: croppedBytes,
      inputImageData: croppedData,
    );
  }
}
*/

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:code_scan/code_scan.dart';

void main() async {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({ Key? key }) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with TickerProviderStateMixin {
  String? code;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: code == null ?
          CodeScanner(
            onScan: (code, details, controller) => setState(() => this.code = code),
            onScanAll: (codes, controller) => print('Codes: ' + codes.map((code) => code.rawValue).toString()),
            formats: const [ BarcodeFormat.qrCode ],
            once: true,
          ) : 
          GestureDetector(
            onTap: () => setState(() => code = null),
            child: Container(
              color: Colors.transparent,
              child: Center(child: Text(code!)),
            ),
          ),
      ),
    );
  }
}

lib/camera_overlay.dart
import 'package:flutter/material.dart';

/// Default overlay displayed on top of the camera
class CodeScannerOverlay extends StatelessWidget {
  final double width;
  final double height;

  const CodeScannerOverlay(this.width, this.height, {Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Center(
          child: Container(
            width: width * 0.8,
            height: 0.8,
            color: Colors.redAccent.withOpacity(0.4),
          ),
        ),
        Center(
          child: Container(
            width: width * 0.8,
            height: width * 0.7,
            decoration: BoxDecoration(
              border: Border.all(color: Colors.black.withOpacity(0.3), width: 0.5),
              borderRadius: BorderRadius.circular(10),
            ),
          ),
        ),
        ColorFiltered(
          colorFilter: ColorFilter.mode(Colors.black.withOpacity(0.2), BlendMode.srcOut),
          child: Stack(
            fit: StackFit.expand,
            children: [
              Container(decoration: BoxDecoration(
                color: Colors.black,
                backgroundBlendMode: BlendMode.dstIn,
              )),
              Center(
                child: Container(
                  width: width * 0.8,
                  height: width * 0.7,
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(10),
                  ),
                ),
              ),
            ]
          ),
        ),
      ],
    );
  }
}

lib/camera_listener.dart
import 'dart:async';

import 'package:camera/camera.dart';
import 'package:code_scan/utils/image.dart';
import 'package:google_mlkit_barcode_scanning/google_mlkit_barcode_scanning.dart';
import 'package:rxdart/rxdart.dart';

/// Create a camera listener to plug it with any camera controller, to scan for codes
class CodeScannerCameraListener {
  final CameraController controller;
  final imageController = StreamController<CameraImage>();
  final BarcodeScanner scanner;
  final bool once;
  
  final void Function(String? code, Barcode details, CodeScannerCameraListener listener)? onScan;
  final void Function(List<Barcode> barcodes, CodeScannerCameraListener listener)? onScanAll;
  final void Function(Object error, CodeScannerCameraListener listener)? onError;

  CodeScannerCameraListener(
    this.controller,
    {
      List<BarcodeFormat> formats = const [ BarcodeFormat.all ],
      Duration interval = const Duration(seconds: 1),
      this.once = false,
      
      this.onScan,
      this.onScanAll,
      this.onError,
    }
  ) : this.scanner = BarcodeScanner(formats: formats) {
    imageController.stream
      .throttleTime(interval, leading: false, trailing: true)
      .listen((image) async {
        try {
          final barcodes = await processImage(image);
          
          if (!controller.value.isStreamingImages || barcodes.isEmpty) return;

          if (once) await dispose();
          onScan?.call(barcodes.first.rawValue, barcodes.first, this);
          onScanAll?.call(barcodes, this);
        } catch (error) {
          if (onError != null) {
            onError?.call(error, this);
          } else {
            rethrow;
          }
        }
      });
      
    controller.startImageStream((CameraImage image) {
      if (!imageController.isClosed) imageController.add(image);
    });
  }

  Future<void> dispose() async {
    if (controller.value.isStreamingImages) await controller.stopImageStream();
    await imageController.close();
    await scanner.close();
  }

  Future<List<Barcode>> processImage(CameraImage image) async =>
    await scanner.processImage(image.toInputImage(controller.description));
}

lib/code_scan.dart
library code_scan;

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:rxdart/rxdart.dart';
import 'package:camera/camera.dart';
import 'package:google_mlkit_barcode_scanning/google_mlkit_barcode_scanning.dart';

import 'package:code_scan/utils/image.dart';
import 'package:code_scan/utils/list.dart';

export 'package:camera/camera.dart' show ResolutionPreset;
export 'package:google_mlkit_barcode_scanning/google_mlkit_barcode_scanning.dart' show BarcodeFormat;

/// # Code Scanner
/// 
/// A flexible code scanner for QR codes, barcodes and many others. Using [Google's ML Kit](https://developers.google.com/ml-kit/vision/barcode-scanning). Use it as a Widget with a camera or use the methods provided, with a camera controller..
/// 
/// ## Features
/// 
/// * Scan Linear and 2D formats: QR Code, Barcode, ...
/// * Widget with integrated camera
/// * Listen for callbacks with every code scanned
/// * Choose which formats to scan
/// * Overlay the camera preview with a custom view
/// * Camera lifecycle
/// 
/// ## Scannable Formats
/// 
/// * Aztec
/// * Codabar
/// * Code 39
/// * Code 93
/// * Code 128
/// * Data Matrix
/// * EAN-8
/// * EAN-13
/// * ITF
/// * PDF417
/// * QR Code
/// * UPC-A
/// * UPC-E
class CodeScanner extends StatefulWidget {
  final CameraController? controller;

  /// Which camera to use:
  /// * front
  /// * back
  /// * external
  /// 
  /// Default: `back`
  final CameraLensDirection direction;

  /// Quality of the camera:
  /// * low
  /// * medium
  /// * high
  /// * very high
  /// * ultra
  /// 
  /// or
  /// 
  /// * max
  /// 
  /// Setting a lower resolution preset may not support scanning features on some devices.
  /// It's recommended to use the highest quality preset available, if performance is not an issue.
  /// 
  /// Default: `high`
  final ResolutionPreset resolution;

  /// List of the scannable formats:
  /// * Aztec
  /// * Codabar
  /// * Code 39
  /// * Code 93
  /// * Code 128
  /// * Data Matrix
  /// * EAN-8
  /// * EAN-13
  /// * ITF
  /// * PDF417
  /// * QR Code
  /// * UPC-A
  /// * UPC-E
  final List<BarcodeFormat> formats;

  /// Duration of delay between scans, to prevent lag.
  final Duration scanInterval;

  /// Whether or not, when a code is scanned, the controller should close and no longer scan.
  /// 
  /// Default: `false`
  final bool once;
  final double? aspectRatio;

  /// Called whenever a controller is created.
  /// 
  /// A new controller is created when initializing the widget and when the life cycle is resumed.
  final void Function(CameraController controller)? onCreated;
  final void Function(String? code, Barcode details, CodeScannerCameraListener listener)? onScan;
  final void Function(List<Barcode> barcodes, CodeScannerCameraListener listener)? onScanAll;
  final void Function(Object error, CodeScannerCameraListener listener)? onError;
  
  /// Called whenever camera access permission is denied by the user.
  /// 
  /// Return `true` to retry, else `false`. Not setting this callback, will automatically never retry.
  /// 
  /// Careful when retrying, this permission could have been rejected automatically, if you keep trying then it will silently spam the permission in a cycle. The `error` given can be useful to check this.
  /// 
  /// Another approach would be to request the permission preemptively, before creating this widget, so it will never be needed to handle it here.
  final bool? Function(CameraException error, CameraController controller)? onAccessDenied;

  /// Widget to show before the cameras are initialized.
  final Widget? loading;

  /// Widget to overlay on top of the camera.
  /// 
  /// Default: `CodeScannerOverlay()`
  final Widget? overlay;
  
  /// # Code Scanner
  /// 
  /// A flexible code scanner for QR codes, barcodes and many others. Using [Google's ML Kit](https://developers.google.com/ml-kit/vision/barcode-scanning). Use it as a Widget with a camera or use the methods provided, with a camera controller.
  /// 
  /// ## Features
  /// 
  /// * Scan Linear and 2D formats: QR Code, Barcode, ...
  /// * Widget with integrated camera
  /// * Listen for callbacks with every code scanned
  /// * Choose which formats to scan
  /// * Overlay the camera preview with a custom view
  /// 
  /// ## Simple Usage
  /// ```dart
  /// CodeScanner(
  ///   onScan: (code, details, controller) => ...,
  ///   formats: [ BarcodeFormat.qrCode ],
  ///   once: true,
  ///   onAccessDenied: (error, controller) {
  ///     Navigator.of(context).pop();
  ///     return false;
  ///   },
  /// )
  /// ```
  const CodeScanner({
    super.key,
    this.controller,
    this.direction = CameraLensDirection.back,
    this.resolution = ResolutionPreset.high,
    this.formats = const [ BarcodeFormat.all ],
    this.scanInterval = const Duration(seconds: 1),
    this.once = false,
    this.aspectRatio,

    this.onCreated,
    this.onScan,
    this.onScanAll,
    this.onAccessDenied,
    this.onError,

    this.loading,
    this.overlay,
  });

  @override
  State<CodeScanner> createState() => _CodeScannerState();
}

class _CodeScannerState extends State<CodeScanner> with WidgetsBindingObserver {
  CameraController? controller;
  CodeScannerCameraListener? listener;

  bool retry = true;
  bool initialized = false;
  bool isInternalController = false;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    _initCameraController();
  }

  @override
  void dispose() {
    listener?.dispose();
    controller?.dispose();
    super.dispose();
  }
  
  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    super.didChangeAppLifecycleState(state);
    if (!isInternalController) return;

    if (state == AppLifecycleState.inactive) {
      final CameraController? cameraController = controller;
      if (cameraController == null || !cameraController.value.isInitialized) return;
      cameraController.dispose();
      if (mounted) setState(() => this.controller = null);
    } else if (state == AppLifecycleState.resumed && retry && initialized && controller == null) {
      initialized = false;
      _initCameraController();
    }
  }

  Future<void> _initCameraController() async {
    final CameraController controller;
    final widgetController = widget.controller;
    
    if (widgetController != null) {
      isInternalController = false;
      controller = widgetController;
    } else {
      isInternalController = true;
      controller = await _createCameraController();
      try {
        await controller.initialize();
      } on CameraException catch (error) {
        controller.dispose();
        retry = widget.onAccessDenied?.call(error, controller) ?? false;
        initialized = true;
        return;
      }
    }

    widget.onCreated?.call(controller);
    setState(() => this.controller = controller);

    if (listener != null) listener!.dispose();
    listener = CodeScannerCameraListener(
      this.controller!,
      onScan: widget.onScan,
      onScanAll: widget.onScanAll,
      formats: widget.formats,
      interval: widget.scanInterval,
      once: widget.once,
    );

    initialized = true;
  }
  
  Future<CameraController> _createCameraController() async {
    final cameras = await availableCameras();
    final camera = cameras.firstWhereOrNull((camera) => camera.lensDirection == widget.direction) ?? cameras.first;
    return CameraController(camera, widget.resolution, enableAudio: false);
  }

  @override
  Widget build(BuildContext context) {
    if (controller == null || !controller!.value.isInitialized) return widget.loading ?? Container();
    return CodeScannerCameraView(
      controller: controller!,
      overlay: widget.overlay,
      aspectRatio: widget.aspectRatio,
    );
  }
}

/// Widget to show a camera with an overlay, this widget tries to expand.
class CodeScannerCameraView extends StatelessWidget {
  final CameraController controller;
  final Widget? overlay;
  final double? aspectRatio;

  const CodeScannerCameraView({
    super.key,
    required this.controller,
    this.overlay,
    this.aspectRatio,
  });
  
  @override
  Widget build(BuildContext context) {
    final cameraAspectRatio = controller.value.aspectRatio;
    return LayoutBuilder(
      builder: (context, constraints) => Stack(
        children: [
          SizedBox(
            width: constraints.maxWidth,
            height: constraints.maxHeight,
            child: FittedBox(
              fit: BoxFit.cover,
              clipBehavior: Clip.hardEdge,
              child: SizedBox(
                width: constraints.maxWidth,
                height: constraints.maxWidth * cameraAspectRatio,
                child: CameraPreview(controller),
              ),
            ),
          ),
          overlay ?? CodeScannerOverlay(constraints.maxWidth, constraints.minHeight),
        ],
      ),
    );
  }
}

/// Create a camera listener to plug it with any camera controller, to scan for codes
class CodeScannerCameraListener {
  final CameraController controller;
  final imageController = StreamController<CameraImage>();
  final BarcodeScanner scanner;
  final bool once;
  
  final void Function(String? code, Barcode details, CodeScannerCameraListener listener)? onScan;
  final void Function(List<Barcode> barcodes, CodeScannerCameraListener listener)? onScanAll;
  final void Function(Object error, CodeScannerCameraListener listener)? onError;

  CodeScannerCameraListener(
    this.controller,
    {
      List<BarcodeFormat> formats = const [ BarcodeFormat.all ],
      Duration interval = const Duration(milliseconds: 500),
      this.once = false,
      
      this.onScan,
      this.onScanAll,
      this.onError,
    }
  ) : this.scanner = BarcodeScanner(formats: formats) {
    start();
    imageController.stream
      .throttleTime(interval, leading: false, trailing: true)
      .listen((image) async {
        try {
          await _onImage(image);
        } catch (error) {
          onError?.call(error, this);
        }
      });
  }

  void start() {
    if (!controller.value.isStreamingImages) {
      controller.startImageStream((CameraImage image) {
        if (!imageController.isClosed) imageController.add(image);
      });
    }
  }

  /// Stops the camera image controller and image stream.
  Future<void> stop() async {
    await imageController.close();
    if (controller.value.isStreamingImages) {
      await controller.stopImageStream();
    } 
  }
  
  /// Alias of the `.stop()` method, for consistency with the flutter environment.
  Future<void> dispose() async => await stop();

  Future<List<Barcode>> _onImage(CameraImage image) async {
    if (!controller.value.isStreamingImages) throw Exception('Camera is not streaming images');
    if (onScan == null && onScanAll == null) throw Exception('No listener');

    /*
    final cropWidth = image.width * 0.5;
    final cropHeight = cropWidth * 0.5;
    final cropX = image.width * 0.25;
    final cropY = image.height * 0.5 - cropHeight / 2;
    */

    final inputImage = image
      //.crop(cropX.toInt(), cropY.toInt(), cropWidth.toInt(), cropHeight.toInt())
      .toInputImage(controller.description);

    final barcodes = await scanner.processImage(inputImage);
    await _onImageProcessed(barcodes);
    return barcodes;
  }

  Future<void> _onImageProcessed(List<Barcode> barcodes) async {
    if (!controller.value.isStreamingImages || barcodes.isEmpty) return;

    if (once) await stop();
    onScan?.call(barcodes.first.rawValue, barcodes.first, this);
    onScanAll?.call(barcodes, this);
  }
}

/// Default overlay displayed on top of the camera
class CodeScannerOverlay extends StatelessWidget {
  final double width;
  final double height;

  const CodeScannerOverlay(this.width, this.height, {Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Center(
          child: Container(
            width: width * 0.8,
            height: 0.8,
            color: Colors.redAccent.withOpacity(0.4),
          ),
        ),
        Center(
          child: Container(
            width: width * 0.8,
            height: width * 0.7,
            decoration: BoxDecoration(
              border: Border.all(color: Colors.black.withOpacity(0.3), width: 0.5),
              borderRadius: BorderRadius.circular(10),
            ),
          ),
        ),
        ColorFiltered(
          colorFilter: ColorFilter.mode(Colors.black.withOpacity(0.2), BlendMode.srcOut),
          child: Stack(
            fit: StackFit.expand,
            children: [
              Container(decoration: BoxDecoration(
                color: Colors.black,
                backgroundBlendMode: BlendMode.dstIn,
              )),
              Center(
                child: Container(
                  width: width * 0.8,
                  height: width * 0.7,
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(10),
                  ),
                ),
              ),
            ]
          ),
        ),
      ],
    );
  }
}